Add tx2_60_gazebo

jediofgever opened the pull request:
Did not see gazebo package for tx2_60, therefore this PR

jediofgever committed:
add tx2_60 gazebo support

jediofgever committed:
remove custom gazebo .world

gavanderhoorn commented on staubli_tx2_60_gazebo/package.xml:
```diff
@@ -0,0 +1,44 @@
+      configuration you intend to use them with.
+    </p>
+  </description>
+  <author>Nils Melchert (Leibniz University Hannover)</author>
```
I don't believe this is correct.

gavanderhoorn commented on staubli_tx2_60_gazebo/package.xml:
```diff
@@ -0,0 +1,44 @@
+
+  <buildtool_depend>catkin</buildtool_depend>
+
+  <exec_depend>gazebo</exec_depend>
```
`gazebo` is not a ROS package, but the name of a system dependency. `gazebo_ros` already depends on it, so this `exec_depend` should not be necessary.

Please remove.

gavanderhoorn commented on staubli_tx2_60_gazebo/package.xml:
```diff
@@ -0,0 +1,44 @@
+  <exec_depend>robot_state_publisher</exec_depend>
+  <exec_depend>staubli_resources</exec_depend>
+  <exec_depend>staubli_tx2_60_support</exec_depend>
+  <build_depend>staubli_tx2_60_support</build_depend>
```
Remove.

gavanderhoorn commented on staubli_tx2_60_gazebo/urdf/tx2_60_macro.xacro:
```diff
@@ -0,0 +1,110 @@
+    <robotNamespace>/${prefix}</robotNamespace>
+    <type>transmission_interface/SimpleTransmission</type>
+    <joint name="${prefix}joint_1">
+      <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
```
We may want to discuss whether using `EffortJointInterface` would be better.

It would mean we'd have to tune pids, but interaction with other Gazebo simulated entities would be more robust and of higher quality (position controlled robots in Gazebo are "infinitely stiff", leading to potentially exploding robots when they come into contact with objects).

jediofgever replied:
Sure, I just followed the existing method in other gazebo packages in this repository. So i did not bother to change that. I will try out  `EffortJointInterface` instead of `PositionJointInterface`

jediofgever replied:
@gavanderhoorn  are you aware of any ros-industrial repos that makes use of  `EffortJointInterface`  ?

gavanderhoorn replied:
Yes, `abb_irb1200_gazebo` (see [here](https://github.com/ros-industrial/abb_experimental/blob/04aa3e515c5672cc216180901c59b39dbde99258/abb_irb1200_gazebo/urdf/irb1200_5_90_macro.xacro)).

There's nothing really to it btw, just replace `PositionJointInterface` with `EffortInterface` and start a different controller.

gavanderhoorn replied:
Yes, `abb_irb1200_gazebo` (see [here](https://github.com/ros-industrial/abb_experimental/blob/04aa3e515c5672cc216180901c59b39dbde99258/abb_irb1200_gazebo/urdf/irb1200_5_90_macro.xacro)).

There's nothing really to it btw, just replace `PositionJointInterface` with `EffortInterface` and start a different controller.

Tweaking the gains is actually what is quite some work. The infrastructure is not that difficult to setup.

gavanderhoorn reviewed (commented):
Thanks for the PR. I've added some in-line comments.

jediofgever committed:
tx2_60_gazebo: handle requested changes in package.xml

gavanderhoorn commented on staubli_tx2_60_gazebo/package.xml:
```diff
@@ -0,0 +1,40 @@
+  <exec_depend>joint_state_controller</exec_depend>
+  <exec_depend>joint_trajectory_controller</exec_depend>
+  <exec_depend>position_controllers</exec_depend>
+
```
This now seems to be missing an `exec_depend` on the TX2-60 support package.

Only the `build_depend` should have been removed, as it's not needed.

gavanderhoorn commented on staubli_tx2_60_gazebo/package.xml:
```diff
@@ -0,0 +1,40 @@
+  <version>0.1.0</version>
+  <description>
+    <p>
+      ROS-Industrial Gazebo support package for the Staubli TX260 (and variants).
```
```suggestion
      ROS-Industrial Gazebo support package for the Staubli TX2-60 (and variants).
```

gavanderhoorn commented on staubli_tx2_60_gazebo/package.xml:
```diff
@@ -0,0 +1,40 @@
+    </p>
+    <p>
+      This package contains the configuration data and launch files required
+      to simulate the Staubli TX60 and the Staubli TX60L manipulator in Gazebo.
```
```suggestion
      to simulate the Staubli TX2-60 manipulator in Gazebo.
```

gavanderhoorn commented on staubli_tx2_60_gazebo/urdf/tx2_60.xacro:
```diff
@@ -0,0 +1,15 @@
+  <joint name="world-base_link-fixed" type="fixed">
+    <parent link="world"/>
+    <child link="base_link"/>
+    <origin xyz="0.0 0.0 0.0" rpy="0.0 0.0 0.0" />
```
```suggestion
    <origin xyz="0 0 0" rpy="0 0 0" />
```

gavanderhoorn reviewed (commented):
Some additional comments.

gavanderhoorn commented:
@jediofgever :bell: ?

jediofgever commented:
@gavanderhoorn sorry , I have been busy with some other project, will look into this in near future

jediofgever committed:
tx2-60: apply suggested changes in PR for tx2-60 gazebo support

jediofgever commented:
I wonder whether it is intentional that, this stack does not have `staubli_resources` package. I suggest we have that, because the packages in this stack all have `exec_depend` on that.

jediofgever committed:
position_controllers -> effort_controllers

jediofgever commented:
welcome to give it a test with [jediofgever/simple_trajectory](https://github.com/jediofgever/simple_trajectory).
I did not try to adjust PID gains as the ones I copied from ABB seemed work OK. 
Edit; hmm this gains aren't that great , will try tune and update here

gavanderhoorn commented:
> I wonder whether it is intentional that, this stack does not have `staubli_resources` package. I suggest we have that, because the packages in this stack all have `exec_depend` on that.

That is because that package is available from `ros-industrial/staubli` and should not be duplicated here.

We cannot merge your PR if it introduces a duplicate package.

gavanderhoorn commented:
I'm going to close due to inactivity.

@jediofgever: I'm really happy you made the effort of submitting a PR and iterating on it, but in its current state it cannot be merged.

At the very released the duplicated `staubli_resources` package needs to be removed.

I'm also not sure what is going on with your fork -- you seem to have copied lots of files in there.

gavanderhoorn commented:
If you have time to come back to this at a later date, please update and re-open.

jediofgever commented:
@gavanderhoorn , tx2_60 owners will be quite sad if this doesn’t gets merged 😄, 

You were right about difficulty of tuning the PIDs for effort controllers. Any chance we go back to the position controllers ? 

Resource package will be removed too

jediofgever commented:
I removed the duplicated package and reverted to position_controllers https://github.com/jediofgever/staubli_experimental/commit/4272905797865a049991c2e96bd90e3109b67f41, at this point `tx2_60_gazebo`  package is equal to other gazebo packages in this repository, however I am not sure whether you will want it to be merged.  

could be tested with  ;
```
git clone https://github.com/jediofgever/simple_trajectory.git
cd simple_trajectory && mkdir build && cd build
cmake .. 
make
./simple_trajectory
```
